### PR TITLE
Added onExplosion Scarpet Event

### PR DESCRIPTION
**Pull request by BisUmTo**
Added `__on_explosion()` event.

I didn't add the event to the `event_test.sc` since there are no other global events (like `__on_lightning()` one).

Example: 
```
__on_explosion(pos, power, source, create_fire, blocks) -> (
    print('');
    print('__on_explosion(pos, power, source, create_fire, blocks)');
	print('an explosion occurred:');
	print('  - pos: '+pos);
	print('  - power: '+power);
	print('  - source: '+source);
	print('  - create_fire: '+create_fire);
	reduced_blocks = slice(blocks, 0, 5);
	print('  - blocks: '+(reduced_blocks += '...'))
)
```

### CODE REVIEW

File: docs/scarpet/Full.md [hunk 1]
```diff
@@ -3995,6 +3995,10 @@ at the same time.
 Triggered right after a lightning strikes. Lightning entity as well as potential horseman trap would 
 already be spawned at that point. `mode` is `true` if the lightning did cause a trap to spawn. 
 
+### `__on_explosion(pos, power, source, create_fire, blocks)`
+Triggered right before an explosion blows up, right before any changes to the world are done, but the decision to blow up is already made.  
+The parameter `blocks` contains the list of block that will blow up (or would blow up if `explosionNoBlockDamage` is setted to `true`).
+
```
Comment (James103) [inline docs/scarpet/Full.md:4001]: This is not needed due to GitHub Actions automatically merging any documentation changes to the topic documentation files into Full.md.

File: docs/scarpet/Full.md [hunk 2]
```diff
@@ -4023,7 +4027,7 @@ Example events that may cause it to happen is releasing a bow. The event is trig
 the request, however the `item_tuple` is provided representing the item that the player started with. You can use that and
 compare with the currently held item for a delta.
 
-### `__on_player_finishes_using_item(player, item_tuple, hand))`
+### `__on_player_finishes_using_item(player, item_tuple, hand)`
```
Comment (James103) [inline docs/scarpet/Full.md:4030]: Not needed, same as above.

File: docs/scarpet/Full.md [hunk 3]
```diff
@@ -4133,7 +4137,7 @@ ingested by the player. The exact position of these items is unknown as technica
 items could be spread all across the inventory.
 
 ### `__on_player_connects(player)`
-Triggered when the player has successfully logged in and was placed in the gaem.
+Triggered when the player has successfully logged in and was placed in the game.
```
Comment (James103) [inline docs/scarpet/Full.md:4140]: Not needed, same as above.

File: docs/scarpet/api/Events.md [hunk 4]
```diff
@@ -58,6 +58,10 @@ at the same time.
 Triggered right after a lightning strikes. Lightning entity as well as potential horseman trap would 
 already be spawned at that point. `mode` is `true` if the lightning did cause a trap to spawn. 
 
+### `__on_explosion(pos, power, source, create_fire, blocks)`
+Triggered right before an explosion blows up, right before any changes to the world are done, but the decision to blow up is already made.  
+The parameter `blocks` contains the list of block that will blow up (or would blow up if `explosionNoBlockDamage` is setted to `true`).
```
Comment (James103) [inline docs/scarpet/api/Events.md:63]:
```suggestion
The parameter `blocks` contains the list of blocks that will blow up (or would blow up if `explosionNoBlockDamage` is set to `true`).
```

**Comment by James103** [comment]
Note: This PR closes #621.

**Comment by Ghoulboy78** [comment]
fixes #621 then ig

File: docs/scarpet/Full.md (hunk 1 above)
Comment (BisUmTo) [reply docs/scarpet/Full.md:4001]: Wow! I've never heard about GitHub Action before!
Nice to know!

**Comment by Ghoulboy78** [comment]
BTW Bisumto, I would suggest separating scarpet mixins from non-scarpet mixins. This is cos gnembon said so, so idk what his verdict here will be, but I think his intention is to have scarpet and fc as separate, for which he also needs mixins to be separate.

**Comment by BisUmTo** [comment]
> BTW Bisumto, I would suggest separating scarpet mixins from non-scarpet mixins. This is cos gnembon said so, so idk what his verdict here will be, but I think his intention is to have scarpet and fc as separate, for which he also needs mixins to be separate.

At first I tried to separate them, but I really don't know how to add the event scheduler between the TNTlogger and the clear of the `setAffectBlocks` list...

**Comment by Ghoulboy78** [comment]
Here are my proposed changes (cos i dont feel like openining pr;)
Create a new file called ExplosionMixin_scarpetEventMixin.java (carpet.mixins)
```java
package carpet.mixins;

import net.minecraft.entity.damage.DamageSource;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraft.world.explosion.Explosion;
import org.spongepowered.asm.mixin.Final;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

import java.util.List;

import static carpet.script.CarpetEventServer.Event.EXPLOSION;

@Mixin(value = Explosion.class)
public abstract class ExplosionMixin_scarpetEventMixin {

    @Shadow @Final private World world;
    @Shadow @Final private double x;
    @Shadow @Final private double y;
    @Shadow @Final private double z;
    @Shadow @Final private float power;
    @Shadow @Final private DamageSource damageSource;
    @Shadow @Final private boolean createFire;
    @Shadow @Final private List<BlockPos> affectedBlocks;

    @Inject(method = "affectWorld", at = @At("HEAD"),
            cancellable = true)
    private void onExplosionB(boolean spawnParticles, CallbackInfo ci)
    {
        if (EXPLOSION.isNeeded()) {
            EXPLOSION.onExplosion((ServerWorld) world, x, y, z, power, damageSource, createFire, affectedBlocks);
        }
    }
}
```
then add it to carpet.mixins.json, and then undo ur changes to ExplosionMixin.java. Idk if it will run tho, so maybe test before u commit, but I dont see why it shouldnt.

**Comment by BisUmTo** [comment]
Is `,cancellable= true` needed?

**Comment by BisUmTo** [comment]
I've aready made this java class, but I can't apply it after the other Mixin (to be able to Inject the code before the list cleats).
I tried to change the order of the functions into the json, but it didn't work to me...

**Comment by BisUmTo** [comment]
If anyone has an idea how to be able to pass the `affectedEntities` list as parameter and get the `affectedBlocks` before `explosionNoBlockDamage` is applied, feels free to contribute.

I think that those 2 could be really cool addition to this event.

File: src/main/java/carpet/script/CarpetEventServer.java [hunk 5]
```diff
@@ -809,6 +809,26 @@ public void onCarpetRuleChanges(ParsedRule<?> rule, ServerCommandSource source)
...
             }
         };
+        public static final Event EXPLOSION = new Event("explosion", 5, true)
+        {
+            @Override
+            public void onExplosion(ServerWorld world, double x, double y, double z, float power, DamageSource source, boolean createFire, List<BlockPos> affectedBlocks)
+            {
+                handler.call(
+                        () -> Arrays.asList(
+                                new ListValue(
```
Comment (gnembon) [inline src/main/java/carpet/script/CarpetEventServer.java:819]: ListValue.fromTriple(x, y, z)

File: src/main/java/carpet/mixins/Explosion_scarpetEventMixin.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import java.util.List;
+
+import static carpet.script.CarpetEventServer.Event.EXPLOSION;
+
+@Mixin(value = Explosion.class)
```
Comment (gnembon) [inline src/main/java/carpet/mixins/Explosion_scarpetEventMixin.java:19]: You might want to apply proper priorities to make sure ExplosionMixin with optimized explosions is mixed in later (so not cancelled the event)

**Comment by gnembon** [comment]
priority works counter-intitively - I think is lower priority - the earlier order. 
Regarding entities - it would be possible but that's a lost cause - then entity damaging code list all potential entities and then evaluates them one by one, and damage them in place, so there is a moment when some entities are already affected and some are not even checked yet.
But you could have mixed in the middle of collectBlocksAndDamageEntities - allowing to trigger the event before it affects entites as well.

File: src/main/java/carpet/mixins/Explosion_scarpetEventMixin.java (hunk 6 above)
Comment (BisUmTo) [reply src/main/java/carpet/mixins/Explosion_scarpetEventMixin.java:19]: Ehm... I think it's the reverse 😂
Higher values first, then lower values.

In this case I have to mix the optimization first and then inject my event handler at the beginning of the method with the injection - I had to give a lower prioritize to the event-mixin.

**Comment by BisUmTo** [comment]
> But you could have mixed in the middle of collectBlocksAndDamageEntities - allowing to trigger the event before it affects entites as well.

Yeah, that could be a way to prevent the "Bedrock-Blockstate" Mixin to...
I don't think I have the "Mixin" skills to do that _[since I wrote my first "INJECT AT" a few minutes ago]_, but tomorrow I might give it a try!

**Comment by gnembon** [comment]
you could mixin at  = @At(value='INVOKE', target = 'getOtherEntities').

This would mean that call to the event needs to be added manually to https://github.com/gnembon/fabric-carpet/blob/master/src/main/java/carpet/helpers/OptimizedExplosion.java#L100

**Comment by gnembon** [comment]
Any updates on moving that mixin to a more targetted place? Or I should just merge it and fiddle with it myself?

**Comment by BisUmTo** [comment]
> Any updates on moving that mixin to a more targetted place? Or I should just merge it and fiddle with it myself?

Until Saturday I will be very busy...
I can work on It this weekend

**Comment by gnembon** [comment]
yeah, there is no rush.

**Comment by gnembon** [comment]
So what we are doing with that @BisUmTo ?

**Comment by BisUmTo** [comment]
I tried to make a mixin that's able to obtain the list of affected blocks ~ 2 weeks ago, but I gave up that weekend. 
Last Saturday I spent the whole day learning things about Mixin, so I hope to be able to create the injection as soon as possible.
I try to bring the code now and try again 😉

**Comment by BisUmTo** [comment]
Code used for tests:
```
__on_explosion(pos, power, source, create_fire, blocks, entities) -> (
    print('');
    print('__on_explosion(pos, power, source, create_fire, blocks)');
	print('an explosion occurred:');
	print('  - pos: '+pos);
	print('  - power: '+power);
	print('  - source: '+source);
	print('  - create_fire: '+create_fire);
	reduced_blocks = slice(blocks, 0, 5);
        if(length(blocks) > 5, reduced_blocks  += '...');
	print('  - blocks: '+reduced_blocks);
	reduced_entities = slice(entities, 0, 5);
        if(length(entities) > 5, reduced_entities += '...');
	print('  - entities: '+reduced_entities)
);
```

**Comment by Ghoulboy78** [comment]
I would recomment you add that to event_test.sc cos its an event tracked by scarpet

**Comment by BisUmTo** [comment]
On `event_test.sc` there are only Player based events at the moment...
`__on_lightning()`, for example, is not present...

**Comment by Ghoulboy78** [comment]
oh ye whoops, my bad

File: src/main/java/carpet/helpers/OptimizedExplosion.java [hunk 7]
```diff
@@ -154,6 +161,11 @@ public static void doExplosionA(Explosion e, ExplosionLogHelper eLogger) {
                             densityCache.put(pair, density);
                         }
 
+                        // If it is needed, it saves the entity
+                        if (EXPLOSION.isNeeded()) {
```
Comment (gnembon) [inline src/main/java/carpet/helpers/OptimizedExplosion.java:165]: I would extract here boolean event = EXPLOSION.isNeeded();
and just use if (event)
```suggestion
                        if (eventNeeded) {
```

File: src/main/java/carpet/helpers/OptimizedExplosion.java [hunk 8]
```diff
@@ -68,9 +70,14 @@
...
     private static BlockPos blastChanceLocation;
 
+    // Creating entity list for scarpet event
+    private static List<Entity> entityList = new ArrayList<>();
+
     public static void doExplosionA(Explosion e, ExplosionLogHelper eLogger) {
         ExplosionAccessor eAccess = (ExplosionAccessor) e;
         
+        entityList.clear();
+
```
Comment (gnembon) [inline src/main/java/carpet/helpers/OptimizedExplosion.java:80]:
```suggestion
boolean eventNeeded = EXPLOSION.isNeeded();
```

File: docs/scarpet/api/Events.md [hunk 9]
```diff
@@ -58,6 +58,11 @@ at the same time.
 Triggered right after a lightning strikes. Lightning entity as well as potential horseman trap would 
 already be spawned at that point. `mode` is `true` if the lightning did cause a trap to spawn. 
 
+### `__on_explosion(pos, power, source, create_fire, blocks, entities)`
+Triggered right before an explosion blows up, right before any changes to the world are done, but the decision to blow up is already made.  
+The parameter `blocks` contains the list of blocks that will blow up (empty if `explosionNoBlockDamage` is set to `true`).
+The parameter `entities` contains the list of entities that will be affected by the explosion.
```
Comment (gnembon) [inline docs/scarpet/api/Events.md:64]:
```suggestion
The parameter `entities` contains the list of entities that have been affected by the explosion.
```

File: docs/scarpet/api/Events.md [hunk 10]
```diff
@@ -58,6 +58,11 @@ at the same time.
 Triggered right after a lightning strikes. Lightning entity as well as potential horseman trap would 
 already be spawned at that point. `mode` is `true` if the lightning did cause a trap to spawn. 
 
+### `__on_explosion(pos, power, source, create_fire, blocks, entities)`
+Triggered right before an explosion blows up, right before any changes to the world are done, but the decision to blow up is already made.
```
Comment (gnembon) [inline docs/scarpet/api/Events.md:62]:
```suggestion
Triggered right before an explosion blows up and any changes to the blocks are done, but the decision to blow up is already made and entities are already affected.  
```

File: src/main/java/carpet/mixins/Explosion_scarpetEventMixin.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        if (EXPLOSION.isNeeded())
+        {
+            affectedEntities = new ArrayList<>();
+        }
+    }
+
+    @Inject(method = "collectBlocksAndDamageEntities", locals= LocalCapture.CAPTURE_FAILHARD, at=@At(value = "INVOKE", target = "Lnet/minecraft/entity/Entity;damage(Lnet/minecraft/entity/damage/DamageSource;F)Z"))
+    private void onExplosion(CallbackInfo ci, Set<BlockPos> s1, float f1, int i1, int i2, int i3, int i4, int i5, int i6, List<Entity> l1, Vec3d v1, int i7, Entity entity)
+    {
+        if (EXPLOSION.isNeeded())
```
Comment (gnembon) [inline src/main/java/carpet/mixins/Explosion_scarpetEventMixin.java:55]:
```suggestion
        if (affectedEntities != null)
```

File: src/main/java/carpet/helpers/OptimizedExplosion.java (hunk 7 above)
Comment (BisUmTo) [reply src/main/java/carpet/helpers/OptimizedExplosion.java]: Is there any benefit? 🤔

**Pull request merged**